refactor: Refactor the client to use zitadel HTTP v2 api

mzaniolo opened the pull request:
A new implementation of the client was made to use HTTP instead of the gRPC.

Completely new api

skomski commented:
I think it would be better to use a two-way approach with code generation as foundation because I don't think it is economic to implement a zitadel client by hand when there is a protobuf definition that we can use to generate most request code automatically. We can build a more usable API on top of that and try better error messages etc but it would be always possible to use the code generated version if the method is not already implemented and some of them maybe only a need a really thin wrapper or none. The WFE uses 30+ methods and the MVP most likely uses even more.

mzaniolo commented on src/users/types.rs:
```diff
@@ -0,0 +1,111 @@
+				.map(|phone| CreateHumanUserPhone { phone: phone.clone(), is_verified: false }),
+			idp_links: value.idp_links(),
+		}
+	}
```
The idea here was to create our own representation of a user. I would like to also use this representation of the user when we search for a user. The `display_name` is build from `first_name` and `last_name`. The `is_verified` is set I'm changing based of the inputs of the `create_human_user` because I thought we don't need this info on the user struct. Do you think it's a good idea separate the internal request from the user representation?

sirewix replied:
It's a generic library, I think we should expose zitadel api types as close as possible to what they really are except cases where the the types are super inconvenient. I don't think we should hide anything, this library can potentially be used outside of famedly where assumptions and requirements are totally different

tlater-famedly replied:
We may need those values internally too.

That said, I appreciate the attempt at making the type more ergonomic for rust purposes, deeply nested structs just to have two attributes linked to a type isn't particularly convenient.

I think the most convenient format for a rust library would be to provide a builder, where optional attributes can be set with function calls, so not everything has to be specified up-front. The attributes should closely match the optionalness of the upstream API, though.

sirewix replied:
[`typed-builder`](https://crates.io/crates/typed-builder)

tlater-famedly commented on src/users/mod.rs:
```diff
@@ -0,0 +1,77 @@
+			idp_id: None,
+		};
+
+		// Note: If phone or email isVerified are not set to true the creation of the
```
At least one auth method must be set up, otherwise Zitadel refuses with the v2 API. As it stands you're not configuring any others, but email/phone verification *is* an auth method.

Zitadel have implemented a way to delay the whole auth method thing with invite links, which we can maybe use to circumvent this: https://github.com/zitadel/zitadel/issues/8310#issue-2410579046

tlater-famedly commented on tests/certs/ca.crt:
```diff
@@ -0,0 +1,20 @@
+-----BEGIN CERTIFICATE-----
```
We should not be committing certificates, even if I would much prefer it if that was possible: https://github.com/famedly/ldap-sync/issues/35#issuecomment-2321730986

mzaniolo replied:
removed

tlater-famedly commented on tests/docker-compose.yaml:
```diff
@@ -0,0 +1,28 @@
+services:
```
You can [set up nextest](https://nexte.st/docs/configuration/setup-scripts/) to do pre-test setup automatically in addition to having a composefile around.

This may also be necessary if you add more than one test to prevent test concurrency from causing issues.

mzaniolo replied:
This I'll fix when I do the tests. I'll make a separate PR just for the tests

tlater-famedly reviewed (commented):
Overall approach looks good, I'd like to see how the stream stuff we discussed pans out, though.

mzaniolo commented:
New approach using generated types from the open api. I can probably write some code to generate most of the functions too. I would like to get some feedback to know if this is a good to way to go before I start clean up the code and addressing the comments.

tlater-famedly commented on src/users/models/add_human_user_request.rs:
```diff
@@ -0,0 +1,250 @@
+use crate::users::models;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct AddHumanUserRequest {
```
Should probably use a macro to reduce the boilerplate here, and give the struct a different name, but looks good otherwise.

tlater-famedly commented on src/users/models/add_human_user_request.rs:
```diff
@@ -0,0 +1,250 @@
+		self.user_id = Some(user_id);
+	}
+
+	pub fn with_user_id(mut self, user_id: String) -> AddHumanUserRequest {
```
It's typical to offer *either* an owned *or* a referenced set of methods, and just name them the attribute they are setting. See e.g. [`reqwest::ClientBuilder`](https://docs.rs/reqwest/latest/reqwest/struct.ClientBuilder.html).

One of the crates that provide macros for this would handle it more nicely anyway, though.

tlater-famedly commented on src/users/models/add_human_user_request.rs:
```diff
@@ -0,0 +1,250 @@
+		&self.email
+	}
+
+	pub fn set_phone(&mut self, phone: models::SetHumanPhone) {
```
If we're providing builders we might as well take a `String` and a `bool` as separate args.

tlater-famedly reviewed (commented):
Assuming this was auto-generated, looks *ok*, but not perfect.

Please also commit the scripts you used to generate the code, and for the future keep the code generation commit separate from the rest so it is easier to review.

mzaniolo commented on src/authentication.rs:
```diff
@@ -0,0 +1,164 @@
+		let client = reqwest::Client::new();
+		let service_account_file = Path::new("tests/zitadel/service-user.json");
+		let token = Token::new(
+			&Url::parse("http://localhost:8080")?,
```
Moved to e2e test. When writing tests I'll make one for testing the renew of the token

mzaniolo commented:
> Assuming this was auto-generated, looks _ok_, but not perfect.
> 
> Please also commit the scripts you used to generate the code, and for the future keep the code generation commit separate from the rest so it is easier to review.
> 
> I caught on to it being auto-generated a bit late, but these are my comments on what it looks like from a public API perspective.

For the test generation I used the [swagger editor](https://editor.swagger.io/) and did some manually clean up of the generated code. Their generation script can be found [here](https://github.com/swagger-api/swagger-codegen)

tlater-famedly commented on .config/starte2e.sh:
```diff
@@ -0,0 +1,28 @@
+while [ $retries -gt 0 ]; do
+	retries=$((retries - 1))
+
+	if curl -s --request GET --url "localhost:8080/debug/ready"; then
```
docker has a built-in `--wait` feature which is probably a better way to achieve this.

tlater-famedly commented on src/v2/mod.rs:
```diff
@@ -0,0 +1,89 @@
+
+/// Format a request for logging purposes; this *must* omit any
+/// secrets or otherwise sensitive information
+fn format_request_for_log(request: &Request) -> String {
```
Sounds sensible, but as @Revanee pointed out, the `http` crate has a feature to hide specific headers, and `reqwest` implements this. Could you check if the header isn't already redacted?

tlater-famedly replied:
That would be best, but it's not necessary for this PR.

tlater-famedly reviewed (dismissed):
Looks pretty good for a start!

emgrav commented:
related: https://github.com/zitadel/zitadel/issues/8686

sirewix commented on src/v2/mod.rs:
```diff
@@ -72,3 +72,18 @@ impl Zitadel {
+		request
+			.headers()
+			.into_iter()
+			.filter(|(name, _)| *name != "authorization")
```
If we want to redact manually we probably want to normalize the header (to lower case for example). Otherwise it can skip `Authorization`

sirewix commented on src/v2/mod.rs:
```diff
@@ -51,7 +51,7 @@ impl Zitadel {
 		let _ = request.timeout_mut().insert(DEFAULT_TIMEOUT);
 
-		tracing::debug!("Request: {request:?}");
+		tracing::debug!("Request: {}", format_request_for_log(&request));
```
`tracing::trace!` maybe?
